File: app/services/query_grammar/wca.py
"""通配碼錨 grammar（#3 收尾）。"""
from __future__ import annotations

import re
from typing import Optional

from app.services.query_tokens import CANTO_CHARS_RE, CODE_TAIL_MIDDLE, is_wildcard_char

_PLUS = re.escape(CODE_TAIL_MIDDLE)


def _tokenize(body: str) -> Optional[list[tuple[str, str]]]:
    tokens: list[tuple[str, str]] = []
    i = 0
    while i < len(body):
        ch = body[i]
        if is_wildcard_char(ch):
            tokens.append(("wild", ch))
            i += 1
        elif ch == CODE_TAIL_MIDDLE:
            tokens.append(("star", ""))
            i += 1
        elif ch.isdigit():
            while i < len(body) and body[i].isdigit():
                tokens.append(("code", body[i]))
                i += 1
        elif CANTO_CHARS_RE.match(ch):
            tokens.append(("ref", ch))
            i += 1
        else:
            return None
    return tokens or None


def _tokens_to_spec(
    tokens: list[tuple[str, str]], *, head_literal: Optional[str] = None
) -> Optional[dict]:
    syllables: list[dict] = []
    if head_literal:
        syllables.append({"literal": head_literal})
    i = 0
    while i < len(tokens):
        kind, val = tokens[i]
        if kind == "wild":
            syllables.append({"wild": True})
            i += 1
        elif kind == "code":
            syllables.append({"code": val})
            i += 1
        elif kind == "star":
            if i + 1 < len(tokens) and tokens[i + 1][0] == "ref":
                syllables.append({"ref": tokens[i + 1][1], "star_before": True})
                i += 2
            else:
                syllables.append({"wild": True})
                i += 1
        elif kind == "ref":
            if syllables and "code" in syllables[-1] and "ref" not in syllables[-1]:
                syllables[-1]["ref"] = val
                i += 1
            else:
                return None
        else:
            return None
    if not syllables:
        return None
    if not any("code" in s for s in syllables) or not any("ref" in s for s in syllables):
        return None
    if head_literal is None and not (tokens and tokens[0][0] == "wild"):
        return None
    slots: list[dict] = []
    for pos, syl in enumerate(syllables):
        if "literal" in syl:
            slots.append({"pos": pos, "kind": "literal_char", "value": syl["literal"]})
        if "code" in syl:
            slots.append({"pos": pos, "kind": "code_digit", "value": syl["code"]})
        if "ref" in syl:
            slots.append({"pos": pos, "kind": "final_anchor", "value": syl["ref"]})
    return {"width": len(syllables), "slots": slots, "head_literal": head_literal}


def parse_wildcard_code_anchor_query(q: str) -> Optional[dict]:
    """通配碼錨：?30人、?30+人、+香?30人（左至右掃描）。"""
    if not q or "@" in q or "=" in q:
        return None
    if re.match(rf"^\d+{_PLUS}", q):
        return None
    m = re.match(rf"^{_PLUS}([一-龥])([?_%][0-9_?%{_PLUS}一-龥]+)$", q)
    if m:
        tokens = _tokenize(m.group(2))
        if not tokens:
            return None
        spec = _tokens_to_spec(tokens, head_literal=m.group(1))
        if spec:
            spec["raw_q"] = q
        return spec
    if q[0] not in "?_%":
        return None
    tokens = _tokenize(q)
    if not tokens:
        return None
    spec = _tokens_to_spec(tokens)
    if spec:
        spec["raw_q"] = q
    return spec


def looks_like_wildcard_code_anchor_query(q: str) -> bool:
    return parse_wildcard_code_anchor_query(q) is not None

File: client/src/db/query/grammar/wca.ts
/** Port of query_grammar/wca. */
import {
  createMatchSpec,
  type ConstraintKind,
  type MatchSpec,
} from '../../position-match/spec.ts';
import { QueryKind } from '../../query-kind.ts';
import type { ParsedQuery, WildcardCodeAnchorQuery } from '../../query-types.ts';
import { isWildcardChar } from './mask.ts';
import { CODE_TAIL_MIDDLE, GRAMMAR_PLUS } from './shared.ts';

function wcaTokenize(body: string): Array<[string, string]> | null {
  const tokens: Array<[string, string]> = [];
  let i = 0;
  while (i < body.length) {
    const ch = body[i]!;
    if (isWildcardChar(ch)) {
      tokens.push(['wild', ch]);
      i += 1;
    } else if (ch === GRAMMAR_PLUS || ch === CODE_TAIL_MIDDLE) {
      tokens.push(['star', '']);
      i += 1;
    } else if (/\d/.test(ch)) {
      while (i < body.length && /\d/.test(body[i]!)) {
        tokens.push(['code', body[i]!]);
        i += 1;
      }
    } else if (/[\u4e00-\u9fff]/.test(ch)) {
      tokens.push(['ref', ch]);
      i += 1;
    } else {
      return null;
    }
  }
  return tokens.length ? tokens : null;
}

function wcaTokensToSpec(
  tokens: Array<[string, string]>,
  headLiteral?: string,
): Omit<WildcardCodeAnchorQuery, 'kind' | 'raw_q'> | null {
  const syllables: Array<Record<string, string | boolean>> = [];
  if (headLiteral) {
    syllables.push({ literal: headLiteral });
  }
  let i = 0;
  while (i < tokens.length) {
    const [kind, val] = tokens[i]!;
    if (kind === 'wild') {
      syllables.push({ wild: true });
      i += 1;
    } else if (kind === 'code') {
      syllables.push({ code: val });
      i += 1;
    } else if (kind === 'star') {
      if (i + 1 < tokens.length && tokens[i + 1]![0] === 'ref') {
        syllables.push({ ref: tokens[i + 1]![1], star_before: true });
        i += 2;
      } else {
        syllables.push({ wild: true });
        i += 1;
      }
    } else if (kind === 'ref') {
      const last = syllables[syllables.length - 1];
      if (last && 'code' in last && !('ref' in last)) {
        last.ref = val;
        i += 1;
      } else {
        return null;
      }
    } else {
      return null;
    }
  }
  if (!syllables.length) {
    return null;
  }
  if (!syllables.some((s) => 'code' in s) || !syllables.some((s) => 'ref' in s)) {
    return null;
  }
  if (!headLiteral && !(tokens[0] && tokens[0][0] === 'wild')) {
    return null;
  }
  const slots: WildcardCodeAnchorQuery['slots'] = [];
  for (let pos = 0; pos < syllables.length; pos++) {
    const syl = syllables[pos]!;
    if ('literal' in syl) {
      slots.push({ pos, kind: 'literal_char', value: String(syl.literal) });
    }
    if ('code' in syl) {
      slots.push({ pos, kind: 'code_digit', value: String(syl.code) });
    }
    if ('ref' in syl) {
      slots.push({ pos, kind: 'final_anchor', value: String(syl.ref) });
    }
  }
  return { width: syllables.length, slots, head_literal: headLiteral };
}

/** Port of wca.parse_wildcard_code_anchor_query */
export function parseWildcardCodeAnchorQuery(q: string): WildcardCodeAnchorQuery | null {
  if (!q || q.includes('@') || q.includes('=')) {
    return null;
  }
  if (/^\d+\+/.test(q)) {
    return null;
  }
  let m = q.match(/^\+([\u4e00-\u9fff])([?_%0-9+\u4e00-\u9fff]+)$/);
  if (m) {
    const tokens = wcaTokenize(m[2]!);
    if (!tokens) {
      return null;
    }
    const spec = wcaTokensToSpec(tokens, m[1]);
    if (!spec) {
      return null;
    }
    return { kind: QueryKind.WILDCARD_CODE_ANCHOR, raw_q: q, ...spec };
  }
  if (!'?_%'.includes(q[0]!)) {
    return null;
  }
  const tokens = wcaTokenize(q);
  if (!tokens) {
    return null;
  }
  const spec = wcaTokensToSpec(tokens);
  if (!spec) {
    return null;
  }
  return { kind: QueryKind.WILDCARD_CODE_ANCHOR, raw_q: q, ...spec };
}

/** Port of query_grammar.wca.to_match_spec */
export function toMatchSpec(parsed: ParsedQuery): MatchSpec | null {
  if (parsed.kind !== QueryKind.WILDCARD_CODE_ANCHOR) {
    return null;
  }
  const q = parsed as WildcardCodeAnchorQuery;
  const spec = createMatchSpec(q.width);
  spec.mask = '?'.repeat(q.width);
  if (!spec.slots) {
    spec.slots = [];
  }
  for (const slot of q.slots) {
    const kind = slot.kind as ConstraintKind;
    spec.slots.push({ pos: slot.pos, kind, value: slot.value });
    if (kind === 'literal_char' && slot.value) {
      spec.mask = spec.mask.slice(0, slot.pos) + slot.value + spec.mask.slice(slot.pos + 1);
    }
  }
  return spec;
}
